core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1839,6 +1844,18 @@ abstract class RDD[T: ClassTag](
   def toJavaRDD() : JavaRDD[T] = {
     new JavaRDD(this)(elementClassTag)
   }
+
+  /**
+   * Whether the RDD is in a barrier stage. Spark must launch all the tasks at the same time for a
+   * barrier stage.
+   *
+   * An RDD is in a barrier stage, if at least one of its parent RDD(s), or itself, are mapped from
+   * a RDDBarrier. This function always returns false for a [[ShuffledRDD]], since a
+   * [[ShuffledRDD]] indicates start of a new stage.
+   */
+  def isBarrier(): Boolean = isBarrier_
+
+  @transient private lazy val isBarrier_ : Boolean = dependencies.exists(_.rdd.isBarrier())
```

Comment:
why do we need a lazy val and a def? can we merge them?

Reply:
This is a performance improvement to cache the value to avoid repeatedly compute `isBarrier()` on a long RDD chain.